parsers/FR_O.py

```diff
@@ -1,351 +1,188 @@
+        "RE": "date_heure",
+        "GF": "date",
+        "MQ": "date_heure",
+        "GP": "date",
+    }
+
+    if target_datetime and zone_key not in HISTORICAL_DATASETS.keys():
+        raise ParserException(
+            "FR_O.py",
+            f"Historical data not implemented for {zone_key} in this parser.",
+            zone_key,
+        )
+    elif target_datetime is None and zone_key not in LIVE_DATASETS.keys():
+        raise ParserException(
+            "FR_O.py",
+            f"Live data not implemented for {zone_key} in this parser.",
+            zone_key,
+        )
+
+    URL_QUERIES: Dict[str, Union[str, None]] = {
+        #   "refine": "statut:Validé" if target_datetime else None,
+        "timezone": "UTC",
+        "order_by": f"{DATE_STRING_MAPPING[zone_key]} desc",
+        "refine": f"{DATE_STRING_MAPPING[zone_key]}:{target_datetime.strftime('%Y')}"
+        if target_datetime
+        else None,
+    }
+
+    url = generate_url(zone_key, target_datetime)
+    response: Response = ses.get(url, params=URL_QUERIES)
+    data = response.json()
+    return data, DATE_STRING_MAPPING[zone_key]
 
 
-def fetch_price(
-    zone_key: str,
+def fetch_production(
+    zone_key: str = "FR-COR",
     session: Optional[Session] = None,
     target_datetime: Optional[datetime] = None,
-    logger: Logger = getLogger(__name__),
+    logger=getLogger(__name__),
 ):
-    if target_datetime is None:
-        raise NotImplementedError("There is no real time data")
-
-    r = session or requests.session()
-    params = get_param(zone_key, target_datetime)
-    api = get_api(zone_key, target_datetime)
-    date_name = get_date_name(zone_key, target_datetime)
-
-    r = session or requests.session()
-
-    # Data retrievement
-    response = r.get(api, params=params)
-    data = json.loads(response.content)
-    datapoints = []
-
-    if len(data["records"]) > 0 and ("fields" in data["records"][0]):
-        for i in range(0, len(data["records"])):
-            fields = data["records"][i]["fields"]
-            if "cout_moyen_de_production_eu_mwh" in fields:
-                datetime_result = arrow.get(fields[date_name]).datetime
-                datapoints.append(
-                    {
-                        "zoneKey": zone_key,
-                        "currency": "EUR",
-                        "datetime": datetime_result,
-                        "source": get_source(zone_key, target_datetime),
-                        "price": float(fields["cout_moyen_de_production_eu_mwh"]),
-                    }
+    data, date_string = fetch_data(zone_key, session, target_datetime, logger)
+    if data == []:
+        raise ParserException(
+            "FR_O.py",
+            f"No data available for {zone_key} for {target_datetime.strftime('%Y')}"
+            if target_datetime
+            else f"No live data available for {zone_key}.",
+            zone_key,
+        )
+    return_list: List[Dict[str, Any]] = []
+    for object in data:
```

Comment:
Maybe improve the naming here?
Data = production_mixes?
object = production_mix?
The idea is that we can understand this line of code straight away without looking at the surrounding context.

Reply:
The object structure looks like this:
```json
[
  {
    "territoire": "Corse",
    "statut": "Estim\u00e9",
    "date_heure": "2021-01-01T00:00:00+00:00",
    "production_totale_mw": 333.2,
    "thermique_mw": 90.5,
    "hydraulique_mw": 86.2,
    "micro_hydraulique_mw": 11.7,
    "photovoltaique_mw": -0.7,
    "eolien_mw": 3.7,
    "bioenergies_mw": 0.8,
    "importations_mw": 141.0,
    "cout_moyen_de_production_eur_mwh": null
  },
  {
    "territoire": "Corse",
    "statut": "Estim\u00e9",
    "date_heure": "2021-01-01T03:00:00+00:00",
    "production_totale_mw": 275.6,
    "thermique_mw": 71.2,
    "hydraulique_mw": 72.8,
    "micro_hydraulique_mw": 11.7,
    "photovoltaique_mw": -1.0,
    "eolien_mw": 2.4,
    "bioenergies_mw": 0.8,
    "importations_mw": 117.7,
    "cout_moyen_de_production_eur_mwh": null
  }
]
So data is the top level array, and object is well the objects in that array. Perhaps adding a comment would enough?

Reply:
In theory it could also return
```json
{
  "message": "ODSQL query is malformed: invalid_function() Clause(s) containing the error(s): select.",
  "error_code": "ODSQLError"
}
```
and
```json
{
  "errorcode": 10002,
  "reset_time": "2021-01-26T00:00:00Z",
  "limit_time_unit": "day",
  "call_limit": 10000,
  "error": "Too many requests on the domain. Please contact the domain administrator."
}
```
I could add specific checks for these.